### PR TITLE
[move-package] Move package cli

### PR DESCRIPTION
**Pull request by tzakian**
This PR builds on PR #9030 and adds CLI support for the package system. It adds support for three basic Move Package commands to the CLI. Currently, these are:
* `move package build`: build the package
* `move package errmap`: build an errmap for the package (and its dependencies)
* `move package prove`: run the Move Prover on the package. Note that the package is currently subject to restrictions in #9030. Prover flags can be passed through this command as well. 

This PR does not add unit testing support, that will be a separate PR. 

In the process of adding support the Move CLI `lib.rs` was getting a bit hard to follow, so I refactored the CLI structure so that each subcommand module holds its CLI args and argument handler. This PR _does not_ convert the CLI over to using the package system, nor does it convert over the move-stdlib and diem-framework. These will be done in upcoming PRs.

### CODE REVIEW

File: language/tools/move-cli/src/experimental/cli.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+#[derive(StructOpt)]
+pub enum ExperimentalCommand {
+    /// Perform a read/write set analysis and print the results for
+    /// `module_file`::`script_name`
+    #[structopt(name = "read-write-set")]
+    ReadWriteSet {
+        /// Path to .mv file containing module bytecode.
+        #[structopt(name = "module")]
+        module_file: String,
```
Comment (modocache) [inline language/tools/move-cli/src/experimental/cli.rs:20]: How about:

```suggestion
        #[structopt(name = "module", parse(from_os_str))]
        module_file: std::path::PathBuf,
```

`analyze_read_write_set` could also be modified to take a `std::path::Path` argument for the module file. But I don't know if this is a meaningful improvement, so feel free to ignore this suggestion.

File: language/tools/move-cli/src/lib.rs [hunk 2]
```diff
@@ -92,6 +88,18 @@ pub struct MoveCLI {
 
 #[derive(StructOpt)]
 pub enum Command {
+    #[structopt(name = "package")]
+    Package {
+        /// Path to package. If none is supplied the current directory will be used.
+        #[structopt(long = "path", short = "p", global = true)]
+        path: Option<String>,
```
Comment (modocache) [inline language/tools/move-cli/src/lib.rs:95]: Same as above, technically this could be a `PathBuf`. Up to you!

(There are several other options in this pull request that use `String` to refer to a file path. I'll stop commenting on each one.)

File: language/tools/move-cli/src/package/cli.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        name: String,
+    },
+    /// Build the package at `path`
+    #[structopt(name = "build")]
+    Build,
+    /// Generate error map for the package and its dependencies at `path`
+    #[structopt(name = "errmap")]
+    ErrMapGen {
+        /// The prefix that all error reasons will be prefixed with, e.g., "E"
+        error_prefix: Option<String>,
```
Comment (modocache) [inline language/tools/move-cli/src/package/cli.rs:34]: Sorry if I'm missing something obvious, but, what use cases do you anticipate for this feature? Maybe examples could be included in the docstring?

File: language/tools/move-cli/src/experimental/cli.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+use move_core_types::{
+    language_storage::TypeTag, parser, transaction_argument::TransactionArgument,
+};
+
+use structopt::{clap::arg_enum, StructOpt};
+
+#[derive(StructOpt)]
+pub enum ExperimentalCommand {
+    /// Perform a read/write set analysis and print the results for
+    /// `module_file`::`script_name`
```
Comment (modocache) [inline language/tools/move-cli/src/experimental/cli.rs:15]: Extremely nit-picky, but just pointing out: some of the option docstrings in this pull request end with a period `.` and some don't. My preference would be to have all of them end with a period -- but please don't waste time on it if you don't mind either way :)

File: language/tools/move-cli/src/sandbox/cli.rs [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        /// exists in global storage
+        #[structopt(long = "no-republish")]
+        no_republish: bool,
+        /// By default, code that might cause breaking changes for bytecode
+        /// linking or data layout compatibility checks will not be published.
+        /// Set this flag to ignore breaking changes checks and publish anyway
+        #[structopt(long = "ignore-breaking-changes")]
+        ignore_breaking_changes: bool,
+        /// fix publishing order
+        #[structopt(short = "m", long = "override-ordering")]
```
Comment (modocache) [inline language/tools/move-cli/src/sandbox/cli.rs:38]: Nit-pick: personally I prefer not to have short option names if there's no clear single letter that makes sense for the option. In this case, `m` doesn't appear in the name `override-ordering`, so I wouldn't make it available as a short name.

File: language/tools/move-cli/src/sandbox/cli.rs [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+        /// If set, fail during compilation when attempting to publish a module that already
+        /// exists in global storage
+        #[structopt(long = "no-republish")]
+        no_republish: bool,
+        /// By default, code that might cause breaking changes for bytecode
+        /// linking or data layout compatibility checks will not be published.
+        /// Set this flag to ignore breaking changes checks and publish anyway
+        #[structopt(long = "ignore-breaking-changes")]
+        ignore_breaking_changes: bool,
+        /// fix publishing order
```
Comment (modocache) [inline language/tools/move-cli/src/sandbox/cli.rs:37]: Could you expand on this docstring? How is the order being fixed, and in what situations might a user wish to specify this option?

**Review by modocache — APPROVED** [review]
LGTM! Just some minor nit-picks, plus a request for additional documentation on one option.

File: language/tools/move-cli/src/package/cli.rs (hunk 3 above)
Comment (tzakian) [reply language/tools/move-cli/src/package/cli.rs:34]: This is used for generating error mappings that you can use with the explanation tool. e.g., instead of `MoveAbort(4096)` you can use this error map to get human readable explanations. I'll add some more to the docstring :)

File: language/tools/move-cli/src/experimental/cli.rs (hunk 4 above)
Comment (tzakian) [reply language/tools/move-cli/src/experimental/cli.rs]: Fixed! As long as we're consistent I'm happy :)

File: language/tools/move-cli/src/lib.rs (hunk 2 above)
Comment (tzakian) [reply language/tools/move-cli/src/lib.rs:95]: Went through and updated things to use `PathBuf` and `Path` where appropriate instead of `String`.

**Review by modocache — DISMISSED** [review]
